### PR TITLE
Explain what aux_data is for when registering custom pytree nodes

### LINKED ISSUE #8726: Doc for register_pytree_node_class doesn't explain aux_data
The documentation [here](https://jax.readthedocs.io/en/latest/pytrees.html?highlight=register_pytree_node_class#extending-pytrees) and [here](https://jax.readthedocs.io/en/latest/_autosummary/jax.tree_util.register_pytree_node_class.html) does not explain what `aux_data` is useful for (`aux_data` is set to `None`). Some more explanations and an example would be welcome.

### PR DESCRIPTION
**Pull request by PhilHippo**
Fixes #8726.

The issue is that `aux_data` is never explained, and every example passes `None` for it, so there is nothing for a reader to learn from. That is still true today in both places the issue points at.

**`jax/_src/tree_util.py`**

`register_pytree_node_class` documents its argument as "a type to register as a pytree" and says nothing about the `tree_flatten` / `tree_unflatten` methods it actually requires, unlike `register_pytree_node` directly above it. This documents both, and explains that auxiliary data is stored in the treedef rather than among the leaves, so it must be hashable and compare by value, and is treated as static by `jax.jit`.

The doctest example now carries a real `name` through `aux_data` instead of `None`, and shows the two consequences: the name does not appear in `jax.tree.leaves`, and two containers with different names have different tree structures.

**`docs/101/pytrees.md`**

The page already explains the split in prose, but `RegisteredSpecial`, `RegisteredSpecial2` and the class-based example all pass `aux_data = None`. This adds one worked example whose auxiliary data holds a label, showing that it stays out of the leaves, that unflattening puts it back, and that it participates in structure comparisons. It links forward to the `meta_fields` discussion already on the page, since that is the same idea with a friendlier API.

Verified locally:

- `pytest --doctest-modules jax/tree_util.py`: 13 passed
- `pytest tests/tree_util_test.py`: 453 passed, 24 skipped
- `ruff check jax/_src/tree_util.py`: clean
- executed every cell of `docs/101/pytrees.md` via jupytext + nbclient with `allow_errors=False`: all pass, and the new cells emit exactly the outputs the surrounding prose claims
- confirmed the re-trace claim separately: calling a jitted function with labels `a`, `a`, `b` traces twice, not three times

Also bumps the freshness `reviewed` date on the guide.

### CODE REVIEW

**Comment by jakevdp** [comment]
Thanks for the contribution – I'm not sure that adding a new section to the docs here is the best: with docs like this, less is often more. That said it's true that we don't have any examples where `aux_data` is not `None`. Maybe it would be better to modify one of the existing examples rather than adding a new section?

**Comment by PhilHippo** [comment]
Fair point, thanks. I dropped the new section entirely and instead made the existing `register_pytree_node_class` example (`RegisteredSpecial2`) carry a `name` through `aux_data`, with a two-sentence lead-in on why it goes there rather than in `children`.

I left the first `RegisteredSpecial` example with `aux_data = None` so the initial flatten/unflatten introduction stays minimal, but happy to move the `name` there instead.

File: jax/_src/tree_util.py [hunk 1]
```diff
@@ -341,7 +341,24 @@ def register_pytree_node_class(cls: Typ) -> Typ:
   a class-oriented interface.
 
   Args:
-    cls: a type to register as a pytree
+    cls: a type to register as a pytree. It must define two methods:
```
Comment (jakevdp) [inline jax/_src/tree_util.py]: I think the reason this docstring is somewhat sparse is that it's always been understood as a lightweight wrapper around `register_pytree_node`, where details like those you add here are more fully described. With that in mind, I think rather than expand this docstring, we should refer to the existing one. We could say something like
```
The class must define a ``tree_flatten`` method and a ``tree_unflatten`` class method,
which implement ``flatten_func`` and ``unflatten_func`` as described in 
:func:`jax.tree_util.register_pytree_node`.
```
What do you think?

**Review by jakevdp — CHANGES_REQUESTED** [review]

**Comment by PhilHippo** [comment]
You're right, it's much cleaner like this - I have made the changes

File: docs/101/pytrees.md [hunk 2]
```diff
@@ -306,27 +306,33 @@ jax.grad(lambda s: s.x ** 2 + s.y)(RegisteredSpecial(3.0, 4.0))
 
 Alternatively, you can define appropriate `tree_flatten` and `tree_unflatten`
 methods on your class and decorate it with
-{func}`~jax.tree_util.register_pytree_node_class`:
+{func}`~jax.tree_util.register_pytree_node_class`. This version also carries a
```
Comment (jakevdp) [inline docs/101/pytrees.md:309]: Sorry for the nitpick, but when I read this it sounds like it's saying `register_pytree_node_class` also carries a `name`.

The original text shows two different ways of doing the same thing – the updated text now shows two different ways of doing two slightly different things. Perhaps it would be clearer if both examples included a static `name` attribute?

Comment (PhilHippo) [reply docs/101/pytrees.md]: Don't worry, both examples now carry a static `name` through `aux_data`.

**Review by jakevdp — APPROVED** [review]
Looks good, thanks!
